rust/lance/src/dataset.rs

```diff
@@ -486,17 +488,63 @@ impl Dataset {
             });
         }
 
-        let fragments = write_fragments(
-            object_store.clone(),
-            &self.base,
-            &schema,
-            stream,
-            params.clone(),
-        )
-        .await?;
+        let transaction = match (params.tail_compaction_size, self.get_fragments().last()) {
+            (Some(tail_compact_thresh), Some(last_frag))
+                if last_frag.count_rows().await? <= tail_compact_thresh =>
+            {
+                let last_data = last_frag.scan().try_into_stream().await?;
+                let last_frag_stream: SendableRecordBatchStream = last_data.into();
+                let arrow_schema = last_frag_stream.schema();
+                let merged_stream = {
+                    // need to scope this use because TokioStreamExt collides with futures::StreamExt
+                    use tokio_stream::StreamExt as TokioStreamExt;
+                    RecordBatchStreamAdapter::new(arrow_schema, last_frag_stream.merge(stream))
+                };
 
-        let transaction =
-            Transaction::new(self.manifest.version, Operation::Append { fragments }, None);
+                let fragments = write_fragments(
+                    object_store.clone(),
+                    &self.base,
+                    &schema,
+                    Box::pin(merged_stream),
+                    params.clone(),
+                )
+                .await?;
+
+                if fragments.len() > 1 {
```

Comment:
Can't we encounter this scenario if:

* The last fragment is small
* The current fragment is large (large enough to be broken into multiple files)

Reply:
Yeah, I think we can remove this limit. I added this because I wanted to reuse `Rewrite` op, if we are adding. If we add the brandnew fragments into a `Rewrite` op, we would have to re-generate all the new fragements on conflict.

I'll add a new op type to be aware of this so to not regenerate the ones that doesn't contain data from the old fragement